LUCENE-10045: Extend ConjunctionUtils to support LongValues and DoubleValues

gsmiller opened the pull request:
# Description

This change introduces the idea of a DISI that can wrap a single "leader" iterator and a list of "followers," exposing the functionality through `ConjunctionUtils`. It then uses it to simplify logic in the facets module (there might be other places we could use this also?).

# Solution

New DISI impl that implements conjunction logic with a lead DISI and a list of "followers".

# Tests

Created new unit tests; existing tests are all still passing as well.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.

gsmiller committed:
LUCENE-10045: Extend ConjunctionUtils to support LongValues and DoubleValues

nocommit: Draft change only for early feedback

msokolov commented on lucene/core/src/java/org/apache/lucene/search/ConjunctionUtils.java:
```diff
@@ -97,4 +97,11 @@ public static void addIterator(
+  /**
+   * Create a conjunction for a given lead {@link DocIdSetIterator} and an array of {@link Values}.
+   */
+  public static DocIdSetIterator createValuesConjunction(DocIdSetIterator lead, Values[] values) {
```
wdyt about using `Values ...` in the signature instead? At least for your one use case it will be cleaner.

gsmiller replied:
Thanks for the suggestion! Definitely more flexible and modern. I need to spend a little more time looking at the rest of the `ConjunctionUtils` use-cases to determine if it makes more sense to use a `List` here for consistency with the rest of the helper methods. Hoping to make some time to dig a little deeper later in this week. Thanks again!

gsmiller replied:
I went ahead and made this a `List` to be consistent with other factory methods in `ConjunctionUtils`

jpountz commented on lucene/core/src/java/org/apache/lucene/search/Values.java:
```diff
@@ -0,0 +1,33 @@
+ * Per-segment, per-document values, which can be calculated at search-time See also: {@link
+ * LongValues}, {@link DoubleValues}
+ */
+public abstract class Values {
```
can we make this class pkg-private?

gsmiller replied:
I don't think so. Since `ConjunctionUtils` is meant to expose conjunction functionality publicly, this needs to be defined outside of the immediate package. Even beyond that, it wouldn't be accessible to the facets module, where I was hoping to more immediately use it.

gsmiller replied:
I've also just realized that the signature for this `Values` class is identical to the existing `DocValuesIterator` in `core/index`, which is currently package-private. Any thoughts on expanding the visibility of `DocValuesIterator` to be public and using it in this change instead? It would expand this functionality to all of the `DocValuesIterator` sub-classes, which might actually be pretty useful.

gsmiller replied:
Ah, never mind! I overlooked the fact that `DocValuesIterator` extends `DocIdSetIterator`, so that doesn't work. The idea is to define an API contract where implementing classes only need to support `advanceExact(int)` (and may not be able to actually define `next()` or `advance()`). Disregard please :)

gsmiller replied:
While thinking about this more though, I _did_ rename it to reflect the more general concept of something that can be positioned on a specified doc but cannot necessarily provide iteration itself. I like this idea in general (and the idea of being able to wrap a "leader" and list of "followers" with a DISI), but I'd be curious if others are opposed.

gsmiller committed:
move from array to list and add an initial unit test

gsmiller committed:
add tests; still probably want some more

gsmiller committed:
more general naming

gsmiller committed:
simplify more facet code using new conjunction util

gsmiller committed:
spotless

gsmiller commented:
I've added some tests and updated faceting code to take advantage of this change. Curious what people thing of the idea! Thanks for the feedback already provided as well!

jpountz commented:
While having abstraction for conjunction of DocIdSetIterators makes sense to me because it encapsulates important logic such as sorting iterators by increasing cost, I'm not convinced of the value of this pull request. It saves one line of code here or there, but at the cost of the additional `DocIdPositionable` API, which we said we wanted to avoid in previous discussions, see e.g. [LUCENE-9081](https://issues.apache.org/jira/browse/LUCENE-9081).

gsmiller commented:
@jpountz I think that's a fair view of this work. Apologies for overlooking the past related discussions as well; looks like it does overlap a fair amount with these.

From my side, I've seen a couple different use-cases in how we use Lucene where we essentially do an inner join across multiple `DoubleValues`/`LongValues` and only do some processing when a doc contains a value in all of them. So this functionality would be a bit more helpful in those cases, but I agree that it doesn't save a substantial amount of code (but would be nice to write it once and reuse it, even though it's not that significant).

Coming back to Lucene more generally though, there aren't as many applications for this as I anticipated, and the applications I've found are "joining" with a single value only, so I think it would be hard to make a case for adding this.

I'll park this change and close the PR for now (and pull this code into a local utility for our specific use-cases). Thanks as always for the feedback!
